### PR TITLE
nadsatt dom-api

### PR DESCRIPTION
**Pull request by nadsatt**
dom-api

[Demo](https://nadsatt.github.io/frontend-2021-homeworks/submissions/nadsatt/dom-api/) |
[Code base](https://github.com/nadsatt/frontend-2021-homeworks/tree/dom-api/submissions/nadsatt/dom-api)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/nadsatt/dom-api/app.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    }
+    else {
+        names.forEach(name => {
+            let sidebarItem = document.createElement('li');
+            sidebarItem.textContent = name;
+            sidebarItem.className = 'sidebar__item';
+
+            sidebarItem.onclick = function(){
+                let id = movies.find(movie => movie.name === this.textContent).id;
+                fillPageByMovieId(id)
```
Comment (madmaxWMFU) [inline submissions/nadsatt/dom-api/app.js:85]: After each click, you go through all the steps again.
Think about how you can avoid this and relieve this problem.

**Review by madmaxWMFU — CHANGES_REQUESTED** [review]
Good job!
See comment below

File: submissions/nadsatt/dom-api/app.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    if(!ratingSectionContent.firstElementChild){
+        const fragment = document.createDocumentFragment();
+        const container = document.createElement('div');
+    
+        for (let i = 1; i <= 10; i++){
+            container.insertAdjacentHTML('beforeend', ratingIcon);
+            container.lastElementChild.classList.add('rating-section__icon');
+        }
+
+        while (container.firstElementChild){
```
Comment (MaxKalashnyk) [inline submissions/nadsatt/dom-api/app.js:113]: Do you really need this `while` here?

File: submissions/nadsatt/dom-api/app.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        const sidebarItems = sidebar.querySelectorAll('.sidebar__item');
+        sidebarItems.forEach((item, i) => item.textContent = names[i]);
+    }
+    else {
+        names.forEach(name => {
+            let sidebarItem = document.createElement('li');
+            sidebarItem.textContent = name;
+            sidebarItem.className = 'sidebar__item';
+
+            sidebarItem.onclick = function(){
```
Comment (MaxKalashnyk) [inline submissions/nadsatt/dom-api/app.js:83]: Use event delegation, please. More info [here](https://learn.javascript.ru/event-delegation)

File: submissions/nadsatt/dom-api/app.js [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+function toggleSidebar(){
+    headerIcon.classList.toggle('icon-wrapper--active');
+    sidebarIcon.classList.toggle('icon-wrapper--active');
+    sidebar.classList.toggle('sidebar--active');
+}
+
+function fillPageByMovieId(id){
+    const movie = movies.find(movie => id === movie.id);
+
+    fillSidebarList(movies.map(movie => movie.name));
```
Comment (MaxKalashnyk) [inline submissions/nadsatt/dom-api/app.js:63]: You can move this function outside. It's no need to call it multiple times.

File: submissions/nadsatt/dom-api/app.js [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    fillPosterSection(movies[id].name);
+    fillRatingSection(movies[id].rating);
+    fillActorsSection(movies[id].name);
+    fillDescriptionSection(movies[id].description);
+    defineColors(movies[id].colors);
+    defineBackgrounds(movies[id].name);
+}
+
+function fillSidebarList(names){
+    if(sidebarList.firstElementChild){
```
Comment (MaxKalashnyk) [inline submissions/nadsatt/dom-api/app.js:73]: When you'll move `fillSidebarList` function outside your `fillPageByMovieId` you can get rid of the checking for first element existence.

**Review by MaxKalashnyk — CHANGES_REQUESTED** [review]
Good job! It's really awesome app!
Look at my comments, please. 
I agree with @madmaxWMFU that you should eliminate calling all these functions on each click. It has a big influence on the performance of your app. Try to think about how you can do it. And let me know if you'll have any issues with that.

File: submissions/nadsatt/dom-api/index.html [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            <div class="sidebar__icon-wrapper icon-wrapper">
+                <div class="sidebar__icon icon"></div>
+            </div>
+        </div>
+        <p class="header__text">catflix</p>
+    </header>
+    <main class="main">
+        <section class="poster-section section">
+            <p class="poster-section__heading heading"></p>
+            <img class="poster-section__content">
```
[the remark this answers is not in the document]
Comment (nadsatt) [reply submissions/nadsatt/dom-api/index.html:31]: `alt` depends on selected movie so i add id dynamically using js

**Comment by nadsatt** [comment]
@madmaxWMFU, @MaxKalashnyk, thanks a lot for such a detailed review! I've opened for myself a bunch of new good practices and fixed everything. Please re-review.

P.S. While inspecting my code according to your notes I even found one more bug (6bc1925).

File: submissions/nadsatt/dom-api/app.js [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    establishSidebarToogling();
+    establishSidebarItemSelection();
+
+    createInitialElements();
+    
+    fillPageByMovieId(0);
+
+    function establishSidebarToogling(){
+        headerIcon.addEventListener('click', toggleSidebar);
+        sidebarIcon.addEventListener('click', toggleSidebar);
```
Comment (madmaxWMFU) [inline submissions/nadsatt/dom-api/app.js:59]: You can add an event handler to DOMContentLoaded without a function

File: submissions/nadsatt/dom-api/app.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        colors: ['rgb(23,131,72)', 'rgb(93, 150, 18)', 'rgb(175, 171, 171)', 'rgb(128,126,44)']
+    },
+    ];
+
+    establishSidebarToogling();
+    establishSidebarItemSelection();
+
+    createInitialElements();
+    
+    fillPageByMovieId(0);
```
Comment (madmaxWMFU) [inline submissions/nadsatt/dom-api/app.js:55]: Add number 0 to variable 
`const initialFilm = 0;`

File: submissions/nadsatt/dom-api/app.js [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+    function establishSidebarToogling(){
+        headerIcon.addEventListener('click', toggleSidebar);
+        sidebarIcon.addEventListener('click', toggleSidebar);
+    }
+
+    function establishSidebarItemSelection(){
+        sidebarList.addEventListener('click', e => {
```
Comment (madmaxWMFU) [inline submissions/nadsatt/dom-api/app.js:63]: Add an event handler to DOMContentLoaded and create the nested function separately

**Review by madmaxWMFU — CHANGES_REQUESTED** [review]

File: submissions/nadsatt/dom-api/app.js (hunk 7 above)
Comment (nadsatt) [reply submissions/nadsatt/dom-api/app.js:59]: @MaxKalashnyk adviced me to use function due to it's flexibility (https://github.com/kottans/frontend-2021-homeworks/pull/82#discussion_r519316780). I didn't understood if it was an advice for future or request for change so i changed it. Should i put the property approach back?

Comment (madmaxWMFU) [reply submissions/nadsatt/dom-api/app.js:59]: No, only write like this
```
document.addEventListener('DOMContentLoaded', ()=>{
    establishSidebarToogling();
    establishSidebarItemSelection();
    createInitialElements();
    headerIcon.addEventListener('click', toggleSidebar);
    sidebarIcon.addEventListener('click', toggleSidebar);
    fillPageByMovieId(0);
});

```

File: submissions/nadsatt/dom-api/app.js (hunk 9 above)
Comment (nadsatt) [reply submissions/nadsatt/dom-api/app.js:63]: Do you mean moving code of `DOMContentLoaded` handler into separate func?

Comment (nadsatt) [reply submissions/nadsatt/dom-api/app.js:63]: or should i add one more `DOMContentLoaded` handler with code of `establishSidebarItemSelection` func?

Comment (madmaxWMFU) [reply submissions/nadsatt/dom-api/app.js:63]: add handler into `DOMContentLoaded ` and write the nested function separately

Comment (nadsatt) [reply submissions/nadsatt/dom-api/app.js]: thanks, i got it and fixed!

it was really strange to create function wrapper not for **handler**, but for **code which assign handler**. I don't even know why i did it 😄

File: submissions/nadsatt/dom-api/app.js [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    },
+];
+
+function toggleSidebar(){
+    headerIcon.classList.toggle('icon-wrapper--active');
+    sidebarIcon.classList.toggle('icon-wrapper--active');
+    sidebar.classList.toggle('sidebar--active');
+}
+
+function selectSidebarItem(e){
```
Comment (MaxKalashnyk) [inline submissions/nadsatt/dom-api/app.js:65]: you can destructure your argument here
```suggestion
function selectSidebarItem({target}){
```

File: submissions/nadsatt/dom-api/app.js [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    const target = e.target;
+
+    if(target instanceof HTMLLIElement){
```
Comment (MaxKalashnyk) [inline submissions/nadsatt/dom-api/app.js:68]: It's better to use `target.classList.contains('sidebar__item')` here

File: submissions/nadsatt/dom-api/app.js [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+        let id = movies.find(movie => movie.name === target.textContent).id;
```
Comment (MaxKalashnyk) [inline submissions/nadsatt/dom-api/app.js:69]: If this `Array.prototype.find()` can return `undefined` if it doesn't find any element by this condition. So it's better to check for the existence of this element and then get the id of it.

File: submissions/nadsatt/dom-api/app.js [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+        sidebarItem.classList.add('sidebar__item')  
+    
+        sidebarList.append(sidebarItem);
+    });
+}
+
+function createRatingIcons(){
+    const fragment = document.createDocumentFragment();
+    const container = document.createElement('div');
+    const iconsNumber = 10;
```
Comment (MaxKalashnyk) [inline submissions/nadsatt/dom-api/app.js:95]: You can make it a global constant at the beginning of your code

File: submissions/nadsatt/dom-api/app.js [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+
+        fragment.appendChild(container.lastElementChild);
+    }
+
+    ratingSectionContent.append(fragment); 
+}
+
+function createActorsImgs(){
+    const fragment = document.createDocumentFragment();
+    const imgsNumber = 4;
```
Comment (MaxKalashnyk) [inline submissions/nadsatt/dom-api/app.js:109]: Let's make it a global constant as well

**Review by MaxKalashnyk — CHANGES_REQUESTED** [review]
Good job! It looks much better!
Look at my comments, please.
And one more important thing I noticed:
Try to go to the Developer tools --> Network --> Throttling --> Set Slow 3G. And notice that your app is loading soooo slow. It's because of the heavy images. They are soooo heavy! I recommend using image compressors. 
[TinyPNG](https://tinypng.com/) - for the size of the images less than 5MB
[Optimizilla](https://imagecompressor.com/)
[Compress JPEG](https://compressjpeg.com/)
And others compressors...
They can help you to increase page loading speed.

P.S. And one more recommendation for the future tasks. No need to add all of these images to your PR. html, css and scripts will be enough.

**Review by MaxKalashnyk — APPROVED** [review]
Perfect! Approved

**Pull request merged**